core/repository/sail/src/main/java/org/eclipse/rdf4j/repository/sail/SailRepositoryConnection.java

```diff
@@ -208,21 +225,30 @@ public SailQuery prepareQuery(QueryLanguage ql, String queryString, String baseU
 	@Override
 	public SailTupleQuery prepareTupleQuery(QueryLanguage ql, String queryString, String baseURI)
 			throws MalformedQueryException {
-		ParsedTupleQuery parsedQuery = QueryParserUtil.parseTupleQuery(ql, queryString, baseURI);
+		Optional<TupleExpr> sailTupleExpr = sailConnection.prepareQuery(ql, Query.Type.TUPLE, queryString, baseURI);
+		ParsedTupleQuery parsedQuery = sailTupleExpr.isPresent()
+				? new ParsedTupleQuery(queryString, sailTupleExpr.get())
+				: QueryParserUtil.parseTupleQuery(ql, queryString, baseURI);
 		return new SailTupleQuery(parsedQuery, this);
```

Comment:
This looks like it ends up parsing only once.  Thats good :)

Btw. Optionals have some rather nice chaining potential in java 8.

```
		ParsedTupleQuery parsedQuery = sailTupleExpr
			.map(expr -> new ParsedTupleQuery(queryString, expr))
			.orElse(QueryParserUtil.parseTupleQuery(ql, queryString, baseURI));
```

Reply:
I always forget about Optional.map for some reason. Good call, I'll fix that.